packages/squiggle-lang/src/public/SqValuePath.ts

```diff
@@ -1,133 +1,257 @@
           }
         }
+        return [];
       }
-      return [];
-    };
+      case "Array": {
+        for (let i = 0; i < ast.elements.length; i++) {
+          const element = ast.elements[i];
+          if (locationContains(element.location, offset)) {
+            return [
+              SqValuePathEdge.fromIndex(i),
+              ...buildRemainingPathEdges(element),
+            ];
+          }
+        }
+        return [];
+      }
+      case "LetStatement": {
+        return [
+          SqValuePathEdge.fromKey(ast.variable.value),
+          ...buildRemainingPathEdges(ast.value),
+        ];
+      }
+      case "DefunStatement": {
+        return [
+          SqValuePathEdge.fromKey(ast.variable.value),
+          ...buildRemainingPathEdges(ast.value),
+        ];
+      }
+      case "Block": {
+        if (
+          ast.statements.length === 1 &&
+          ["Array", "Dict"].includes(ast.statements[0].type)
+        ) {
+          return buildRemainingPathEdges(ast.statements[0]);
+        }
+      }
+    }
+    return [];
+  }
+  return buildRemainingPathEdges(ast);
+}
+
+export class SqValuePath {
+  public root: ValuePathRoot;
+  public edges: SqValuePathEdge[];
 
+  constructor(props: { root: ValuePathRoot; edges: SqValuePathEdge[] }) {
+    this.root = props.root;
+    this.edges = props.edges;
+  }
+
+  static findByAstOffset({
+    ast,
+    offset,
+  }: {
+    ast: ASTNode;
+    offset: number;
+  }): SqValuePath | undefined {
     return new SqValuePath({
       root: "bindings", // not important, will probably be removed soon
-      items: findLoop(ast),
+      edges: astOffsetToPathEdges(ast, offset),
+    });
+  }
+
+  isRoot() {
+    return this.edges.length === 0;
+  }
+
+  lastItem(): SqValuePathEdge | undefined {
+    return this.edges[this.edges.length - 1];
+  }
+
+  extend(edge: SqValuePathEdge) {
+    return new SqValuePath({
+      root: this.root,
+      edges: [...this.edges, edge],
     });
   }
 
-  itemsAsValuePaths({ includeRoot = false }) {
+  uid(): string {
+    return `${this.root}--${this.edges.map((edge) => edge.uid()).join("--")}`;
+  }
+
+  // Checks if this SqValuePath completely contains all of the nodes in this other one.
+  contains(smallerItem: SqValuePath) {
+    if (this.root !== smallerItem.root) {
+      return false;
+    }
+    if (this.edges.length < smallerItem.edges.length) {
+      return false;
+    }
+    for (let i = 0; i < smallerItem.edges.length; i++) {
+      if (!this.edges[i].isEqual(smallerItem.edges[i])) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  isEqual(other: SqValuePath) {
+    if (this.root !== other.root) {
+      return false;
+    }
+    if (this.edges.length !== other.edges.length) {
+      return false;
+    }
+    for (let i = 0; i < this.edges.length; i++) {
+      if (!this.edges[i].isEqual(other.edges[i])) {
+        return false;
+      }
+    }
+    return true;
+  }
+
+  allSqValuePathSubsets({ includeRoot = false }) {
```

Comment:
> isn't mentioning SqValuePath excessive

It's mentioning ``SqValuePath`` because it's returning SqValuePaths. I think that can be easy to forget. 

I like ``prefix`` instead of ``subset``, that's clearly better.

Reply:
Changed to ``allPrefixPaths``